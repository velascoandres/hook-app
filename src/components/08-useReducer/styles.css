body{
    padding: 50px;
}

.list-group-item{
    align-items: center;
    display: flex;
    justify-content: space-between;
    cursor: pointer;
}

.complete {
    text-decoration: line-through;
    color: green;
    transform: scale(1.1,1.1);
    transition: 1s;
}

.uncomplete {
    transition: 1s;
}