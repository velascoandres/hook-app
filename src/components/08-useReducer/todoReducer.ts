import { Reducer } from "react";

export interface ITodo {
    id: number;
    desc: string;
    done: boolean;
};

export type TodosState = ITodo[];

export const initialState: TodosState = [
    {
        id: new Date().getTime(),
        desc: 'Aprender React',
        done: false,
    }
];

export enum TodoActionEnum {
    agregar = 'AGREGAR',
    borrar = 'BORRAR',
    completar = 'COMPLETAR,',
    inicial = 'INICIO',
}

export interface TodoAction {
    type: TodoActionEnum;
};

export interface CrearTodo extends TodoAction {
    type: TodoActionEnum.agregar;
    payload: ITodo;
}

export interface BorrarTodo extends TodoAction {
    type: TodoActionEnum.borrar;
    payload: { id: number };
}

export interface CompletarTodo extends TodoAction {
    type: TodoActionEnum.completar;
    payload: { id: number };
}





export const todoReducer: Reducer<TodosState, TodoAction> = (state: TodosState = [], action: TodoAction) => {
    switch (action?.type) {
        case TodoActionEnum.agregar:
            return [
                ...state,
                (action as CrearTodo).payload,
            ];
        case TodoActionEnum.borrar:
            return state.filter(
                todo => todo.id !== (action as BorrarTodo).payload.id,
            );
        case TodoActionEnum.completar:
            return state.map(
                todo => {
                    const esTodo = todo.id === (action as CompletarTodo).payload.id;
                    if (esTodo) {
                        todo.done = !todo.done;
                        return {
                            ...todo,
                            done: !todo.done,
                        };
                    }
                    return todo;
                },
            );
        default:
            return state;
    }

}