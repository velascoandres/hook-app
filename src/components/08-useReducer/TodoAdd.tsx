import React from 'react'
import { Fragment } from 'react'
import { useForm } from '../../hooks/useForm';
import { ITodo } from './todoReducer';

export type TodoAppProps = {
    handleAddTodo: (todo: ITodo) => void;
};

export const TodoAdd = ({handleAddTodo}: TodoAppProps) => {

    const [{ desc: description }, handleInputChange, reset] = useForm<ITodo>({ id: 0, desc: '', done: false });

    const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
        e.preventDefault();

        if (description.trim().length <= 1) {
            return;
        }

        const newTodo: ITodo = {
            id: new Date().getTime(),
            desc: description,
            done: false,
        };
        handleAddTodo(newTodo);
        reset();
    };

    return (
        <Fragment>
            <h4>Agregar Todo</h4>
            <hr />

            <form onSubmit={handleSubmit}>
                <input
                    type="text"
                    name="desc"
                    placeholder="Aprender ..."
                    autoComplete="off"
                    className="form-control"
                    value={description}
                    onChange={handleInputChange}
                />
                <button
                    type="submit"
                    className="btn btn-outline-primary mt-1 btn-block"
                >
                    Agregar
                </button>
            </form>
        </Fragment>
    );
}
