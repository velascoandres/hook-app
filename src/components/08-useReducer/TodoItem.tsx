import { ITodo } from './todoReducer';


export type TodoItemProps = {
    todo: ITodo;
    handleDelete: (id: number) => () => void;
    handleComplete: (id: number) => () => void;
    index: number;
};

export const TodoItem = ({ todo, index, handleDelete, handleComplete }: TodoItemProps) => {

    const { done, id, desc } = todo;

    return (
        <li
            key={id}
            className="list-group-item"
        >
            <p
                className={done ? "complete" : "uncomplete"}
                onClick={handleComplete(id)}
            >
                {index + 1}. {desc}
            </p>
            <button
                className="btn btn-danger"
                onClick={handleDelete(id)}
            >
                Borrar
            </button>
        </li>
    );
}
