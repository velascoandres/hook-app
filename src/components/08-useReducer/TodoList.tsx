import { TodoItem } from './TodoItem';
import { ITodo } from './todoReducer';

export type TodoListProps = {
    todos: ITodo[];
    handleDelete: (id: number) => () => void;
    handleComplete: (id: number) => () => void;
};

export const TodoList = ({ todos, handleComplete, handleDelete }: TodoListProps) => {
    return (
        <ul className="list-group list-group-flush">
            {
                todos.map(
                    (todo, indice: number) => (
                        <TodoItem
                            todo={todo}
                            index={indice}
                            handleComplete={handleComplete}
                            handleDelete={handleDelete}
                        />
                    ),
                )
            }
        </ul>
    )
}
