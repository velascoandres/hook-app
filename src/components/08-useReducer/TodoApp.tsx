import { useEffect, useReducer } from 'react';

import {
    BorrarTodo,
    CompletarTodo,
    CrearTodo,
    ITodo,
    TodoActionEnum,
    todoReducer,
} from './todoReducer';
import { TodoList } from './TodoList';
import { TodoAdd } from './TodoAdd';

import './styles.css';

const init = () => {

    const rawTodos = localStorage.getItem('todos');

    return rawTodos ? JSON.parse(rawTodos) : [];
};


export const TodoApp = () => {

    // const [todos, dispatch] = useReducer(todoReducer, initialState);
    const [todos, dispatch] = useReducer(todoReducer, [], init);


    useEffect(() => {
        localStorage.setItem('todos', JSON.stringify(todos));
    }, [todos]);


    const handleDelete = (id: number) => {
        return () => {
            const action: BorrarTodo = {
                type: TodoActionEnum.borrar,
                payload: { id, },
            };
            dispatch(action);
        };
    };

    const handleComplete = (id: number) => {
        return () => {
            const action: CompletarTodo = {
                type: TodoActionEnum.completar,
                payload: { id, },
            };
            dispatch(action);
        };
    };

    const handleAddTodo = (newTodo: ITodo) => {
        const action: CrearTodo = {
            type: TodoActionEnum.agregar,
            payload: newTodo,
        };

        dispatch(action);

    }


    return (
        <div>
            <h1>TodoApp ({todos.length})</h1>
            <hr />
            <div className="row">
                <div className="col-7">
                    <TodoList
                        todos={todos}
                        handleComplete={handleComplete}
                        handleDelete={handleDelete}
                    />
                </div>
                <div className="col-5">
                    <TodoAdd
                        handleAddTodo={handleAddTodo}
                    />
                </div>
            </div>
        </div>
    );
}
