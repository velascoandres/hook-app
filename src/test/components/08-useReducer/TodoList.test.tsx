import { shallow } from "enzyme"
import { TodoList } from "../../../components/08-useReducer/TodoList";
import { DEMO_TODOS } from "../../fixtures/demoTodos";

describe('Pruebas en <TodoList />', () => {

    const handleDelete = jest.fn();
    const handleComplete = jest.fn();

    const wrapper = shallow(
        <TodoList
            todos={DEMO_TODOS}
            handleComplete={handleComplete}
            handleDelete={handleDelete}
        />
    );

    test('debe de mostrarse correctamente', () => {

        expect(wrapper).toMatchSnapshot();

    });

    test('debe de tener dos <TodoItem />', () => {
        const items = wrapper.find('TodoItem');
        expect(items.length).toBe(DEMO_TODOS.length);
        expect(items.at(0).prop('handleDelete')).toEqual(
            expect.any(Function),
        );
    });

});
