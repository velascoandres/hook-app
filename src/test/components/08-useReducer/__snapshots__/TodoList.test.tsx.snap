// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoList /> debe de mostrarse correctamente 1`] = `
<ul
  className="list-group list-group-flush"
>
  <TodoItem
    handleComplete={[MockFunction]}
    handleDelete={[MockFunction]}
    index={0}
    key="1"
    todo={
      Object {
        "desc": "Aprender React",
        "done": false,
        "id": 1,
      }
    }
  />
  <TodoItem
    handleComplete={[MockFunction]}
    handleDelete={[MockFunction]}
    index={1}
    key="2"
    todo={
      Object {
        "desc": "Aprender Mongo",
        "done": false,
        "id": 2,
      }
    }
  />
</ul>
`;
